hefquin-engine/src/main/java/se/liu/ida/hefquin/engine/queryplan/utils/LogicalToPhysicalOpConverter.java

```diff
@@ -24,125 +13,32 @@
+		.register( new PhysicalOpBindJoin.Factory() )              // Bind-join for brTPF interface
+		.register( new PhysicalOpBindJoinWithBoundJoin.Factory() ) // Bind-join for SPARQL interface
+		.register( new PhysicalOpBindJoinWithUNION.Factory() )     // (fallback) if no non-joining var available
+		// .register( new PhysicalOpBindJoinWithVALUESorFILTER.Factory() )
+		// .register( new PhysicalOpBindJoinWithFILTER.Factory() )
+		// .register( new PhysicalOpBindJoinWithVALUES.Factory() )
+		.register( new PhysicalOpSymmetricHashJoin.Factory() )     // Inner join
+		.register( new PhysicalOpHashRJoin.Factory() )             // Right outer join
+		.register( new PhysicalOpIndexNestedLoopsJoin.Factory() )  // Index NLJ algorithm, fm to request join partners
+		// .register( new PhysicalOpHashJoin.Factory() )
+		// .register( new PhysicalOpNaiveNestedLoopsJoin.Factory() )
+	;
 
-	public static BinaryPhysicalOp convert( final BinaryLogicalOp lop ) {
-		if (      lop instanceof LogicalOpJoin )     return convert( (LogicalOpJoin) lop );
-		else if ( lop instanceof LogicalOpUnion )    return convert( (LogicalOpUnion) lop );
-		else if ( lop instanceof LogicalOpRightJoin ) return convert( (LogicalOpRightJoin) lop );
-		else throw new UnsupportedOperationException("Unsupported type of logical operator: " + lop.getClass().getName() + ".");
-	}
-
-	public static BinaryPhysicalOp convert( final LogicalOpJoin lop ) {
-		return new PhysicalOpSymmetricHashJoin(lop);
-	}
-
-	public static BinaryPhysicalOp convert( final LogicalOpUnion lop ) {
-		return new PhysicalOpBinaryUnion(lop);
-	}
-
-	public static BinaryPhysicalOp convert( final LogicalOpRightJoin lop ) {
-		return new PhysicalOpHashRJoin(lop);
-	}
-
-	// --------- n-ary operators -----------
-
-	public static NaryPhysicalOp convert( final NaryLogicalOp lop ) {
-		if (      lop instanceof LogicalOpMultiwayJoin )  return convert( (LogicalOpMultiwayJoin) lop );
-		else if ( lop instanceof LogicalOpMultiwayLeftJoin ) return convert( (LogicalOpMultiwayLeftJoin) lop );
-		else if ( lop instanceof LogicalOpMultiwayUnion ) return convert( (LogicalOpMultiwayUnion) lop );
-		else throw new UnsupportedOperationException("Unsupported type of logical operator: " + lop.getClass().getName() + ".");
-	}
-
-	public static NaryPhysicalOp convert( final LogicalOpMultiwayJoin lop ) {
-		throw new UnsupportedOperationException();
-	}
-
-	public static NaryPhysicalOp convert( final LogicalOpMultiwayLeftJoin lop ) {
-		throw new UnsupportedOperationException();
+	public static PhysicalOperator convert( final LogicalOperator lop ) {
+		return convert(lop, null);
 	}
 
-	public static NaryPhysicalOp convert( final LogicalOpMultiwayUnion lop ) {
-		return new PhysicalOpMultiwayUnion();
+	public static PhysicalOperator convert( final LogicalOperator lop, final ExpectedVariables inputVars ) {
+		return registry.create(lop, inputVars);
```

Comment:
With this change, we will have to do a cast in almost every place where we call this `convert` function, like you have done now, for instance, in your changes to `PhysicalPlanFactory`. I would prefer to keep these casts local within `LogicalToPhysicalOpConverter`. To this end, you can replace each of the two `convert` functions here by four different functions:
* NullaryPhysicalOp convert( final NullaryLogicalOp lop )
* UnaryPhysicalOp convert( final UnaryLogicalOp lop )
* BinaryPhysicalOp convert( final BinaryLogicalOp lop )
* NaryPhysicalOp convert( final NaryLogicalOp lop )

... and likewise for the version with `ExpectedVariables` as second argument.

Reply:
Resolved in commit 5d9ca9d2fc6ac9939dd0e699252ed0817014c60a